src/Microsoft.ML.Core/Prediction/TrainContext.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Runtime.Data;
+
+namespace Microsoft.ML.Runtime
+{
+    /// <summary>
+    /// Holds information relevant to trainers. Instances of this class are meant to be constructed and passed
+    /// into <see cref="ITrainer{TPredictor}.Train(TrainContext)"/> or <see cref="ITrainer.Train(TrainContext)"/>.
+    /// This holds at least a training set, as well as optioonally a predictor.
+    /// </summary>
+    public sealed class TrainContext
+    {
+        /// <summary>
+        /// The training set. Cannot be <c>null</c>.
+        /// </summary>
+        public RoleMappedData TrainingSet { get; }
+
+        /// <summary>
+        /// The validation set. Can be <c>null</c>. Note that passing a non-<c>null</c> validation set into
+        /// a trainer that does not support validation sets should not be considered an error condition. It
```

Comment:
>an error condition [](start = 85, length = 18)

Can we encourage trainer implementer to show at least warning?

Reply:
I'd prefer to not. Warnings are already printed by the tooling utilizing it as we see [here for validation sets](https://github.com/TomFinley/machinelearning/blob/eb5d0eec0bfc3801c7bc927891c06c8d401e008d/src/Microsoft.ML.Data/Commands/TrainCommand.cs#L166-L168) and [here for initial predictors](https://github.com/TomFinley/machinelearning/blob/eb5d0eec0bfc3801c7bc927891c06c8d401e008d/src/Microsoft.ML.Data/Commands/TrainCommand.cs#L261-L264). To decentralize the checking, in addition to the problems of having to insert that boilerplate somehow, also runs somewhat contrary to the philosophy behind `TrainContext` which is that we can add new members to that without trainer implementors having to care. Under this proposal, suddenly they'd be encouraged to care, and I'd rather keep that object as simple as possible. Plus of course if we did that things that work through those aforementioned pathways would show two warnings.

---
In reply to: [202148448](https://github.com/dotnet/machinelearning/pull/522#discussion_r202148448) [](ancestors = 202148448)